### DISCUSSION TITLE
error: 'struct in_pktinfo’ has no member named ‘ipi_spec_dst’

### DISCUSSION THREAD

**Discussion by MarcusQSousa** [openssl/openssl#23469] [Q&A]
Im trying to install it on windows, but im getting this error when trying to build. I've tried some dumb things but im new at this, and couldnt find any discussions about this.

error=>
```
crypto/bio/bss_dgram.c: In function ‘pack_local’:
crypto/bio/bss_dgram.c:1178:13: error: ‘struct in_pktinfo’ has no member named ‘ipi_spec_dst’
 1178 |         info->ipi_spec_dst      = local->s_in.sin_addr;
      |             ^~
```

**Comment by hlandau** [comment]
This implies that `OPENSSL_SYS_WINDOWS` isn't defined, which is odd. What platform are you targeting? Are you possibly trying to build for cygwin or similar?

What compiler/build environment are you using, and what Configure arguments are you using? Can you post the output of `perl configdata.pm --dump`?

**Comment by tobi-ostfalia** [comment]
hello, im having the exact same error. i cant run the perl output for reasons im not sure about, but i can tell you about the other stuff.
im using OpenSSL 3.2.1 and targeting mingw-conan-Debug-Windows-x86_64-gcc-13.
GCC is version 13.1.0. 
configure arguments (compiler flags) are as follows:  
compiler: gcc -m64 -Wall -O3 -DL_ENDIAN -DOPENSSL_PIC -DUNICODE -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_MT -DOPENSSL_BUILDING_OPENSSL -DZLIB -DZLIB_SHARED -DNDEBUG -DOPENSSLBIN="\"/ucrt64/bin\""

hope this helps.

**Reply by hlandau** [reply]
Building on Windows requires `OPENSSL_SYS_WINDOWS` to be defined, however this should normally be defined automatically in `include/openssl/e_os2.h`. This ultimately gets defined if `_WIN32` or `_WIN64` is defined.

Microsoft expressly documents these macros as being predefined here: https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170

Of course, you are using mingw. However this post seems to indicate that mingw should normally predefine _WIN32: https://stackoverflow.com/questions/9025708/mingw-not-defining-win32-error-in-preprocessor-directives

`gcc -E -dM empty_file.c` can be used to get a list of compiler predefs, so this may to help determine what is going on here. As a stopgap, adding `-D_WIN32` or `-D_WIN64` as appropriate when calling `Configure` might help?

**Reply by tobi-ostfalia** [reply]
in my case, OpenSSL is being built through conan2. so as far as i know, im unable to add any more flags when calling configure, as its being calles with preset flags (and i havent found a way to change this yet...).
here is the configure call at least: 
```
RUN: perl ./Configure "mingw-conan-Debug-Windows-x86_64-gcc-13" shared --prefix=/ --libdir=lib --openssldir="/c/users/<user>/.conan2/p/b/opens8dc36cd331c6e/p/res" no-unit-test threads PERL=perl no-tests --debug enable-fips no-md2 -DOPENSSL_TLS_SECURITY_LEVEL=2 zlib --with-zlib-include="/c/users/<user>/.conan2/p/b/zlibd3ea2317bd63a/p/include" --with-zlib-lib="/c/users/<user>/.conan2/p/b/zlibd3ea2317bd63a/p/lib"
```

as for `gcc -E -dM empty_file.c`, here are the defines that are either W32 or WIN32, i couldnt find anything containing W64 or WIN64:
```#define __WIN32__ 1
#define _WIN32 1
#define WIN32 1
#define __WIN32 1
#define __MINGW32__ 1
```
isnt it quite strange that nothing containing W64 or WIN64 is defined? especially when the OpenSSL target platform is mingw64.

running the same command in the MSYS2 MINGW64 environment/terminal does yield the WIN64/W64 defines. not entirely sure what to make of this right now though.

**Reply by hlandau** [reply]
`_WIN32` is defined even on 64-bit Windows. Yeah, even the name of the API has a legacy component to it...

The above seems incongruous with what is reported. If you modified bss_dgram.c like so I wonder what would happen:

```c
#ifndef OPENSSL_SYS_WINDOWS
# error not windows
#endif
```

**Reply by tobi-ostfalia** [reply]
Finally managed to get into the openssl source code used in conan, shouldve been a lot easier but oh well...
adding the code, the "not windows" error does get thrown. so it seems OPENSSL_SYS_WINDOWS is truly not being defined.
also, some corrections from my other (latest) post:
the `gcc -E -dM ...` command DOES output the WIN64/W64 defines, that was a mistake on my part. 

ive also managed to change the Configure call and add the -D_WIN64 arg, but that is just causing more errors right now:
```
In file included from include/openssl/x509.h:41,
                 from include/openssl/pem.h:23,
                 from include/openssl/ui.h:30,
                 from include/openssl/engine.h:30,
                 from apps/lib/apps.c:21:
include/openssl/sha.h:111:16: error: expected ':', ',', ';', '}' or '__attribute__' before 'h'
  111 |     SHA_LONG64 h[8];
      |                ^
```

but still going to look into this more. seems strange.

**Reply by moucha19** [reply]
I have the exact same issue with the comparable setup - when the Conan + MSYS2 tries to build OpenSSL which is one of the dependencies of OpenCV. Based on information in this thread, I defined `_WIN32` and now I'm at the same error message, complaining about `SHA_LONG64 h[8];`. Did you manage to resolve it?

**Reply by tobi-ostfalia** [reply]
unfortunately not. i only managed to fix it by switching to Ubuntu 22.04 - and even then i had some problems, but a completely clean VM of it managed to do the trick. good luck!

**Reply by trivediravi13** [reply]
I had similar issue but was able to fix it by tweaking conan profile a bit. Hope this would `help`

`
[settings]
os=Windows
arch=x86_64
compiler=gcc
compiler.version=14
compiler.libcxx=libstdc++11
compiler.cppstd=gnu17
compiler.threads=posix
compiler.exception=seh
build_type=Release

[buildenv]
PATH+=(path)C:\msys64\ucrt64\bin

[runenv]
PATH+=(path)C:\msys64\ucrt64\bin

[conf]
tools.microsoft.bash:subsystem=msys2
tools.microsoft.bash:path=C:\msys64\usr\bin\bash.exe
tools.build:compiler_executables={"c": "gcc", "cpp": "g++"}
`

**Reply by pmalessa** [reply]
I broke your changes to the conan profile down to only needing the 2 conf lines:
```
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=14
os=Windows

[conf]
tools.microsoft.bash:subsystem=msys2
tools.microsoft.bash:path=C:\msys64\usr\bin\bash.exe
```
